include/pybind11/pytypes.h

```diff
@@ -377,20 +380,49 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        }
     }
 
     error_already_set(const error_already_set &) = default;
     error_already_set(error_already_set &&) = default;
 
     inline ~error_already_set() override;
 
+    const char *what() const noexcept override {
+        if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
+            } catch (const std::exception &e) {
+                m_lazy_what
+                    = std::string(
+                          "Unknown internal error occurred while constructing error_string: ")
+                      + e.what();
+                return m_lazy_what.c_str();
+            } catch (...) {
+                m_lazy_what = "Unknown internal error occurred";
```

Comment:
I think it's better to kill the process here. See other comment. If this happens the program needs to be fixed. Generously continuing is likely to do more harm than good.

Reply:
@rwgk Are you suggesting I just remove the try / catch block? Since this function is noexcept, it would just call std::terminate if any exception is actually raised.

Reply:
I really wouldn't mind removing the try catches as the only real exceptions I can think of would be stuff like memory allocations etc.